akka-persistence/src/main/scala/akka/persistence/Persistent.scala

```diff
@@ -246,3 +251,26 @@ private[persistence] final case class PersistentImpl(
     s"PersistentRepr($persistenceId,$sequenceNr,$writerUuid,$timestamp,$metadata)"
   }
 }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object CompositeMetadata {
+  def extract[M: ClassTag](metadata: Option[Any]): Option[M] = {
+    val metadataType = implicitly[ClassTag[M]].runtimeClass
+    metadata.flatMap {
+      case CompositeMetadata(entries) =>
+        entries.collectFirst {
+          case m: M @unchecked if metadataType.isAssignableFrom(m.getClass) => m
+        }
+      case other: M @unchecked if metadataType.isAssignableFrom(other.getClass) => Some(other)
```

Comment:
Is the `assignableFrom` flexibility important? I would have thought exactly this class to be enough (and then the entries could be a map and this a map lookup instead).

Reply:
true, CompositeMetadata can probably be a Map with class keys